File: app/src/main/java/com/lasercyber/lws/ui/common/upgrade/BundledFirmwareVersionGate.java
package com.lasercyber.lws.ui.common.upgrade;

import com.lasercyber.lws.ui.common.utils.UpgradeFileReaderUtils;

import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Pure version gating for APK-bundled control-card firmware filenames.
 */
public final class BundledFirmwareVersionGate {

    public static final String FIRMWARE_FILE_PATTERN = "LSW01H\\d{4}S\\d{4}\\.bin";
    private static final Pattern PATTERN = Pattern.compile(FIRMWARE_FILE_PATTERN, Pattern.CASE_INSENSITIVE);

    private BundledFirmwareVersionGate() {
    }

    public static boolean isValidFirmwareFileName(String fileName) {
        return fileName != null && PATTERN.matcher(fileName).matches();
    }

    /**
     * @return true when hardware matches and bundled software version is strictly greater.
     */
    public static boolean isUpgradeCandidate(String bundledFileName, Integer deviceHw, Integer deviceSw) {
        if (!isValidFirmwareFileName(bundledFileName)) {
            return false;
        }
        Integer bundledHw = UpgradeFileReaderUtils.getFileHardwareVersion(bundledFileName);
        Integer bundledSw = UpgradeFileReaderUtils.getFileSoftwareVersion(bundledFileName);
        if (bundledHw == null || bundledSw == null || deviceHw == null || deviceSw == null) {
            return false;
        }
        if (!Objects.equals(bundledHw, deviceHw)) {
            return false;
        }
        return bundledSw > deviceSw;
    }
}

File: app/lws_hmi/lib/features/bundled_firmware/domain/bundled_firmware_version_gate.dart
/// Pure filename gate for APK/App-bundled control-board firmware (lws-ui parity).
abstract final class BundledFirmwareVersionGate {
  /// `LSW01H####S####.bin` (case-insensitive).
  static final RegExp filePattern = RegExp(
    r'^LSW01H\d{4}S\d{4}\.bin$',
    caseSensitive: false,
  );

  static bool isValidFirmwareFileName(String? fileName) {
    if (fileName == null || fileName.isEmpty) {
      return false;
    }
    return filePattern.hasMatch(fileName);
  }

  /// Hardware version from filename (`H` + four digits).
  static int? hardwareVersion(String fileName) {
    if (!isValidFirmwareFileName(fileName)) {
      return null;
    }
    return int.tryParse(fileName.substring(6, 10));
  }

  /// Software version from filename (`S` + four digits).
  static int? softwareVersion(String fileName) {
    if (!isValidFirmwareFileName(fileName)) {
      return null;
    }
    return int.tryParse(fileName.substring(11, 15));
  }

  /// True when HW matches and bundled SW is strictly greater than device SW.
  static bool isUpgradeCandidate({
    required String bundledFileName,
    required int? deviceHw,
    required int? deviceSw,
  }) {
    if (!isValidFirmwareFileName(bundledFileName)) {
      return false;
    }
    final bundledHw = hardwareVersion(bundledFileName);
    final bundledSw = softwareVersion(bundledFileName);
    if (bundledHw == null ||
        bundledSw == null ||
        deviceHw == null ||
        deviceSw == null) {
      return false;
    }
    if (bundledHw != deviceHw) {
      return false;
    }
    return bundledSw > deviceSw;
  }
}
